lib/fs/filesystem.go

```diff
@@ -251,3 +251,18 @@ func Canonicalize(file string) (string, error) {
 
 	return file, nil
 }
+
+// SameFile reports whether fi1 and fi2 describe the same file.
+// For example, on Unix this means that the device and inode fields
+// of the two underlying structures are identical; on other systems
+// the decision may be based on the path names.
+// SameFile only applies to results returned by this package's Stat.
+// It returns false in other cases.
+func SameFile(fi1, fi2 FileInfo) bool {
```

Comment:
The problem I was solving was that the information to compare the files is already available to the core os.SameFile function, but not reachable from my code. If I did that the way you are suggesting, I would have to re-read the inode info for every file/directory and duplicate many code from the core libraries. This would slow down the fs.Lstat which may have a negative effect on Syncthing processing of big directory trees as fs.LStat is called very often. And IMO the duplicated code would make the code less maintainable.

If you read the comment at the fs.SameFile function, it was copied from the os.SameFile function. Even their function works solely for their filesystem implementation. Hence I didi it the same way. Actually, I don't think we need to solve the recursion for other filesystems - in this case fs.SameFile always returns false.

Would it be sufficient for you if I changed the function name from SameFile to sameFile?

Reply:
The needed information is already returned by the os.Lstat (at least for Unix, I didn't check for Windows). My main problem with this approach is that the idea in this package is that the filesystem implementation knows how the specifics works, and provides methods for working with that. This inverts that relationship, having a package level function reach into implementation specifics. The effect is that implementation a new filesystem provider isn't just a matter of doing that implementation, you also need to find this function and add stuff here. That's not great imho.

Reply:
The inode information isn't returned by Lstat on Windows, even though it's available there.
I understand your concern, but there is a trade-off - either make the code clean but unnecessarily complex, or simple but with dependency on implementation instead of just on interface. I chose the second option, because it seems to more maintainable.

But IMO it's not that bad. The function SameFile has a fallback behaviour. So if a new filesystem implementation is created, it needn't be modified. Because, most probably, the new filesystem wouldn't allow infinite recursion.

Reply:
I am not sure I understand why this function can't be moved in to the filesystem?

Literally, replace
```
func SameFile(a, b File) {
}
```
with 
```
func (basicFs) SameFile(a, b File) {
}
```
It does not use anything from the filesystem struct itself, just puts the implementation details behind the filesystem, rather than some global function that needs to know about different variations of filesystems.

Ideally though, I would prefer the Identity() option, and would probably do the extra legwork to read it on windows when it's not available, just because I think it's generally useful piece of information to expose.

Reply:
@AudriusButkevicius "why this function can't be moved in to the filesystem?" fs.SameFile uses fs.FileInfo, BasicFilesystem knows nothing about this interface. At second, walkfs.Walk can be used for various filesystems, not just the BasicFilesystem. It would have to have a different implementation for other filesystems. IMO this is not necessary.

"I think it's generally useful piece of information to expose." What for. If we were writing a library for others to use, I would agree, but what use would it be for in Syncthing? As I've written above, I don't thing it's a good idea. But if you insist on that, I would re-implement that.

Reply:
Walk should just call to fs.SameFile, then every implementation of a filesystem would be responsible for determining if it's the same file or not, regardless of which filesystem the walk is being performed for.

As for Identity(), it could be used for things like better rename detection.
There is a PR proposing to do based on content, but it could be done based on Identity()

Reply:
In general, I understand what you mean, but not in this specific case. SameFile is called within walkFileSystem, which knows nothing about basicFilesystem. How would it call the function? Moreover, SameFile is used from within a (ancestorsDirList) structure, that has access to FileInfo only, but not to the filesystem. I would have to introduce the dependency there.

Reply:
Note that os.SameFile returns false when it's unable to convert the files to os.fileStat. Even when it is actually the same file!

Reply:
It does not need to know about basicFilesystem, it knows about Filesystem which is embedded inside of it.

And filesystem already has SameFile: https://github.com/syncthing/syncthing/blob/master/lib/fs/filesystem.go#L50

Which is already implemented in filesystems where this is relevant:
https://github.com/syncthing/syncthing/blob/master/lib/fs/basicfs.go#L291

So all you need to do in Walk, is call f.SameFile, same way we call f.Lstat:
https://github.com/syncthing/syncthing/blob/master/lib/fs/walkfs.go#L65

calling through to the underlying implementation (because walkfs does not provide one itself, it gains it from the embedded one)

Reply:
I didn't noticed there's SameFile in the basicFilesystem. You are right, this is the right place to put the implementation to. I'll change its implementation and remove the one from Filesystem.

Reply:
SameFile fixed.